src/components/Input.tsx

```diff
@@ -1,22 +1,69 @@
-import { ChangeEventHandler } from 'react';
+import { FormContext } from './Form';
+import { useContext, useState } from 'react';
+import { differenceInYears, isDate } from 'date-fns';
 
-export type InputType = 'email' | 'number' | 'password' | 'text';
+export enum InputType {
+  EMAIL = 'email',
+  PASSWORD = 'password',
+  PHONE = 'tel',
+  BIRTHDATE = 'date',
+  TEXT = 'text',
+}
+
+const regexes: Record<InputType, (input: string) => boolean> = {
+  [InputType.EMAIL]: (input: string) => /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/.test(input),
+  [InputType.PASSWORD]: (input: string) => /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/.test(input),
+  [InputType.TEXT]: (input: string) => input.length > 0,
+  [InputType.PHONE]: (input: string) =>
+    /^(?:(?:\+|00)?(55)\s?)?(?:\(?([1-9][0-9])\)?\s?)?(?:((?:9\d|[2-9])\d{3})-?(\d{4}))$/.test(input),
+  [InputType.BIRTHDATE]: (input: string) => {
+    const birth = new Date(input);
+    const MINIMUM_AGE = 18;
+
+    return isDate(birth) ? differenceInYears(Date.now(), birth) >= MINIMUM_AGE : false;
+  },
+};
 
 interface InputProps {
   id: string;
   labelText: string;
   inputType: InputType;
+  errorMessage: string;
   required?: boolean;
-  error?: string;
-  onChange?: ChangeEventHandler<HTMLInputElement>;
 }
 
 export function Input(props: InputProps) {
+  const [input, setInput] = useState('');
+  const [error, setError] = useState('');
+
+  const formContext = useContext(FormContext);
+
+  const [inputStates, setInputStates] = formContext.inputStates;
+
+  function onChange(e: React.ChangeEvent<HTMLInputElement>) {
+    setError('');
+    setInput(e.target.value);
+
+    setInputStates({ ...inputStates, [props.id]: e.target.value });
```

Comment:
não é uma boa usar o próprio estado dentro do seu set, pode causar atualizações estranhas
```suggestion
    setInputStates((prevState) => ({ ...prevState, [props.id]: e.target.value }));
```

Reply:
Fiz de outra maneira, o que acha?